File: mobile/test/sessionInfo.test.ts
// 会话身份与能力摘要的取数（AR05 §6.1 / §6.3）。
//
// 核心是**四种失败不许混成一种**：明确认证拒绝 / 旧服务端 / 网络未知 / 只取到一半。
// 混成一种的代价很具体：把超时判成 token 失效，用户会在信号不好的地方反复重配
// 一份完全正确的配置。
import {
  fetchSessionInfo,
  parseSessionSummary,
  sessionInfoUrl,
  summaryStale,
} from '@/core/api/sessionInfo'

const EDGE = 'https://vehicle.example.ts.net:8443'
const TOKEN = 'tok-secret'

function fakeFetch(res: Partial<Response> & { json?: () => Promise<unknown> }) {
  const calls: Array<{ url: string; init?: RequestInit }> = []
  const impl = (async (url: string, init?: RequestInit) => {
    calls.push({ url, init })
    return { ok: true, status: 200, json: async () => ({}), ...res } as Response
  }) as unknown as typeof fetch
  return { impl, calls }
}

test('token 只走 Authorization 头，不进 URL（URL 会进访问日志）', async () => {
  const { impl, calls } = fakeFetch({})
  await fetchSessionInfo(EDGE, TOKEN, impl)

  expect(calls[0].url).toBe(sessionInfoUrl(EDGE))
  expect(calls[0].url).not.toContain(TOKEN)
  expect((calls[0].init?.headers as Record<string, string>).Authorization).toBe(`Bearer ${TOKEN}`)
})

test('401 / 403 = 服务端明确拒绝这个 token', async () => {
  for (const status of [401, 403]) {
    const { impl } = fakeFetch({ ok: false, status })
    expect(await fetchSessionInfo(EDGE, TOKEN, impl)).toEqual({ kind: 'unauthorized' })
  }
})

test('404 = 旧服务端，不是鉴权失败', async () => {
  const { impl } = fakeFetch({ ok: false, status: 404 })
  expect(await fetchSessionInfo(EDGE, TOKEN, impl)).toEqual({ kind: 'legacy' })
})

test('网络错误/超时 = 未知，绝不判成 token 失效', async () => {
  const impl = (async () => { throw new Error('Network request failed') }) as unknown as typeof fetch
  const r = await fetchSessionInfo(EDGE, TOKEN, impl)
  expect(r.kind).toBe('unknown')
  expect(r.kind === 'unknown' && r.reason).toContain('Network request failed')
})

test('503 也是未知，不是拒绝', async () => {
  const { impl } = fakeFetch({ ok: false, status: 503 })
  const r = await fetchSessionInfo(EDGE, TOKEN, impl)
  expect(r.kind).toBe('unknown')
})

test('没配服务器时直接给未知，不发请求', async () => {
  const { impl, calls } = fakeFetch({})
  expect((await fetchSessionInfo('', TOKEN, impl)).kind).toBe('unknown')
  expect(calls).toHaveLength(0)
})

test('解析：三种能力状态原样保留，未知状态不被改写成可用', () => {
  const s = parseSessionSummary({
    contract_version: 'ar05.1',
    authenticated: true,
    user_id: 'u1',
    vehicle_id: 'v1',
    authorization_source: 'token',
    granted_scopes: ['vehicle.control', 123],
    capabilities: [
      { id: 'edge-vehicle', display_name: '车辆控制', status: 'available', reason_code: '' },
      { id: 'edge-media', status: 'unauthorized', reason_code: 'scope_missing' },
      { id: 'reminder', status: 'unknown', reason_code: 'query_failed' },
      { status: 'available' },
    ],
    generated_at_ms: 100,
    expires_at_ms: 200,
    summary_status: 'partial',
    summary_reason: 'cloud_unreachable',
  })

  expect(s.grantedScopes).toEqual(['vehicle.control'])
  expect(s.capabilities.map((c) => `${c.id}:${c.status}`)).toEqual([
    'edge-vehicle:available', 'edge-media:unauthorized', 'reminder:unknown',
  ])
  // 没有 id 的行无法指认，丢弃
  expect(s.capabilities).toHaveLength(3)
  // 没给 display_name 就回落 id，不编名字
  expect(s.capabilities[1].displayName).toBe('edge-media')
  expect(s.summaryStatus).toBe('partial')
  expect(s.summaryReason).toBe('cloud_unreachable')
})

test('解析：旧服务端没有 summary_status 时按 complete 处理', () => {
  expect(parseSessionSummary({ user_id: 'u1' }).summaryStatus).toBe('complete')
})

test('解析：PoC 默认放行与 token 授权是两个来源', () => {
  expect(parseSessionSummary({ authorization_source: 'poc_default' }).authorizationSource)
    .toBe('poc_default')
  expect(parseSessionSummary({ authorization_source: 'token' }).authorizationSource).toBe('token')
})

test('摘要过期后标待刷新；没有有效期就不判过期', () => {
  const base = parseSessionSummary({ generated_at_ms: 100, expires_at_ms: 200 })
  expect(summaryStale(base, 199)).toBe(false)
  expect(summaryStale(base, 200)).toBe(true)
  expect(summaryStale(parseSessionSummary({}), 10_000)).toBe(false)
})

File: gateway/edge/session_info_test.go
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"log"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
	"time"

	orchpb "github.com/cockpit/car-agent/gen/go/cockpit/orchestrator/v1"
	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
)

// fakeOrch 只实现 DescribeSession；Handle 不该被这条路径碰到——**只读查询走不到编排**
// 是 AR05 §6.1 的前提（零业务副作用），所以它被调到就是缺陷，直接让测试炸。
type fakeOrch struct {
	orchpb.EdgeOrchestratorClient
	seenMeta map[string]string
	seenCtx  *orchpb.SessionInfoRequest
	resp     *orchpb.SessionInfoResponse
	err      error
}

func (f *fakeOrch) DescribeSession(
	_ context.Context, in *orchpb.SessionInfoRequest, _ ...grpc.CallOption,
) (*orchpb.SessionInfoResponse, error) {
	f.seenCtx = in
	f.seenMeta = in.GetMeta()
	if f.err != nil {
		return nil, f.err
	}
	return f.resp, nil
}

func (f *fakeOrch) Handle(
	context.Context, *orchpb.HandleRequest, ...grpc.CallOption,
) (grpc.ServerStreamingClient[orchpb.HandleEvent], error) {
	panic("session info must not go through the orchestration path")
}

func getSession(t *testing.T, a authConfig, orch orchpb.EdgeOrchestratorClient,
	target string, header map[string]string) *httptest.ResponseRecorder {
	t.Helper()
	request := httptest.NewRequest(http.MethodGet, target, nil)
	for k, v := range header {
		request.Header.Set(k, v)
	}
	response := httptest.NewRecorder()
	handleSessionInfo(response, request, orch, a)
	return response
}

func decodeBody(t *testing.T, r *httptest.ResponseRecorder) map[string]any {
	t.Helper()
	var body map[string]any
	if err := json.Unmarshal(r.Body.Bytes(), &body); err != nil {
		t.Fatalf("response is not JSON: %v (%s)", err, r.Body.String())
	}
	return body
}

func okSummary() *orchpb.SessionInfoResponse {
	return &orchpb.SessionInfoResponse{
		ContractVersion:     "ar05.1",
		Authenticated:       true,
		UserId:              "u1",
		VehicleId:           "v1",
		AuthorizationSource: "token",
		GrantedScopes:       []string{"vehicle.control"},
		Capabilities: []*orchpb.CapabilityStatus{
			{Id: "edge-vehicle", DisplayName: "车辆控制", Status: "available"},
			{Id: "edge-media", Status: "unauthorized", ReasonCode: "scope_missing"},
		},
		GeneratedAtMs:  100,
		ExpiresAtMs:    60_100,
		SummaryStatus:  "complete",
	}
}

func tokenAuth() authConfig {
	return authConfig{
		tokens: map[string]identity{
			"good": {userID: "u1", vehicleID: "v1", scopes: "vehicle.control"},
		},
		defaultUserID:  "u1",
		defaultVehicle: "v1",
		now:            time.Now,
	}
}

func TestSessionInfoStampsScopesFromTheTokenNotTheClient(t *testing.T) {
	orch := &fakeOrch{resp: okSummary()}
	// 客户端在查询串里塞了自己的 scope——网关必须按 token 覆写（同 stampScopes 的唯一权威）
	r := getSession(t, tokenAuth(), orch, "/api/session?token=good&granted_scopes=admin", nil)

	if r.Code != http.StatusOK {
		t.Fatalf("want 200, got %d", r.Code)
	}
	if got := orch.seenMeta["granted_scopes"]; got != "vehicle.control" {
		t.Fatalf("scopes must come from the token, got %q", got)
	}
	if orch.seenMeta["authenticated"] != "true" {
		t.Fatalf("authenticated flag missing: %#v", orch.seenMeta)
	}
	if orch.seenCtx.GetContext().GetUserId() != "u1" {
		t.Fatalf("identity not forwarded: %#v", orch.seenCtx.GetContext())
	}
}

func TestSessionInfoAcceptsBearerHeaderSoTheTokenStaysOutOfTheURL(t *testing.T) {
	orch := &fakeOrch{resp: okSummary()}
	r := getSession(t, tokenAuth(), orch, "/api/session",
		map[string]string{"Authorization": "Bearer good"})

	if r.Code != http.StatusOK {
		t.Fatalf("want 200 for bearer auth, got %d", r.Code)
	}
	if orch.seenMeta["granted_scopes"] != "vehicle.control" {
		t.Fatalf("bearer token was not resolved: %#v", orch.seenMeta)
	}
}

func TestSessionInfoNeverEchoesTheToken(t *testing.T) {
	var logs bytes.Buffer
	previous := log.Writer()
	log.SetOutput(&logs)
	defer log.SetOutput(previous)

	orch := &fakeOrch{resp: okSummary()}
	r := getSession(t, tokenAuth(), orch, "/api/session?token=good", nil)

	if strings.Contains(r.Body.String(), "good") {
		t.Fatalf("token leaked into the response: %s", r.Body.String())
	}
	if strings.Contains(logs.String(), "token=good") {
		t.Fatalf("token leaked into logs: %s", logs.String())
	}
}

func TestSessionInfoRejectsBadTokenOnlyWhenAuthIsRequired(t *testing.T) {
	a := tokenAuth()
	a.required = true
	orch := &fakeOrch{resp: okSummary()}

	r := getSession(t, a, orch, "/api/session?token=nope", nil)
	if r.Code != http.StatusUnauthorized {
		t.Fatalf("want 401 under AUTH_REQUIRED, got %d", r.Code)
	}
	if body := decodeBody(t, r); body["authenticated"] != false {
		t.Fatalf("401 body must say so plainly: %#v", body)
	}

	// 默认模式（AUTH_REQUIRED=false）仍匿名放行，逐字保持既有 WS 行为
	anon := getSession(t, tokenAuth(), orch, "/api/session?token=nope", nil)
	if anon.Code != http.StatusOK {
		t.Fatalf("anonymous fallback must stay, got %d", anon.Code)
	}
	if _, present := orch.seenMeta["granted_scopes"]; present {
		t.Fatalf("anonymous request must not carry forged scopes: %#v", orch.seenMeta)
	}
	if orch.seenMeta["authenticated"] == "true" {
		t.Fatal("anonymous fallback must not claim it authenticated")
	}
}

func TestSessionInfoBackendFailureIsUnknownNotUnauthorized(t *testing.T) {
	// 后端不可达**不是** token 失效：判错方向会让用户去重配一份完全正确的配置。
	orch := &fakeOrch{err: status.Error(codes.Unavailable, "orchestrator down")}
	r := getSession(t, tokenAuth(), orch, "/api/session?token=good", nil)

	if r.Code != http.StatusServiceUnavailable {
		t.Fatalf("want 503 for a dead backend, got %d", r.Code)
	}
	body := decodeBody(t, r)
	if body["summary_status"] != "partial" {
		t.Fatalf("must not fabricate a complete summary: %#v", body)
	}
	if _, present := body["capabilities"]; present {
		t.Fatalf("must not fabricate capabilities: %#v", body)
	}
}

func TestSessionInfoPassesThroughEveryContractField(t *testing.T) {
	orch := &fakeOrch{resp: okSummary()}
	body := decodeBody(t, getSession(t, tokenAuth(), orch, "/api/session?token=good", nil))

	for key, want := range map[string]any{
		"contract_version":     "ar05.1",
		"authenticated":        true,
		"user_id":              "u1",
		"vehicle_id":           "v1",
		"authorization_source": "token",
		"summary_status":       "complete",
	} {
		if body[key] != want {
			t.Fatalf("%s: want %v, got %v", key, want, body[key])
		}
	}
	caps, ok := body["capabilities"].([]any)
	if !ok || len(caps) != 2 {
		t.Fatalf("capabilities not passed through: %#v", body["capabilities"])
	}
	second, _ := caps[1].(map[string]any)
	if second["status"] != "unauthorized" || second["reason_code"] != "scope_missing" {
		t.Fatalf("capability status was rewritten: %#v", second)
	}
	// 空数组必须是 []，不是 null——「一个都没有」与「字段缺失」是两种语义
	if scopes, ok := body["granted_scopes"].([]any); !ok || len(scopes) != 1 {
		t.Fatalf("granted_scopes shape: %#v", body["granted_scopes"])
	}
}

func TestSessionInfoRejectsNonGet(t *testing.T) {
	orch := &fakeOrch{resp: okSummary()}
	request := httptest.NewRequest(http.MethodPost, "/api/session?token=good", nil)
	response := httptest.NewRecorder()
	handleSessionInfo(response, request, orch, tokenAuth())
	if response.Code != http.StatusMethodNotAllowed {
		t.Fatalf("want 405, got %d", response.Code)
	}
}
